### PR TITLE
Bump onyx to fix replay effect (after revert)

### PR DESCRIPTION
**Pull request by chrispader**
@neil-marcellini @roryabraham  @mountiny 

### Details

Original bump PR: https://github.com/Expensify/App/pull/33554

Bumps Onyx to the latest version and fixes the "replay" effect problem with https://github.com/Expensify/react-native-onyx/pull/437.

Bumps Node and NPM versions of Onyx based on https://github.com/Expensify/App/issues/12775#issuecomment-1899331180

Includes changes needed to use Onyx after the recent [TS setup PR](https://github.com/Expensify/react-native-onyx/pull/452)

### Fixed Issues

For https://github.com/Expensify/App/issues/12775
PROPOSAL:

### Tests

- [x] Verify that no errors appear in the JS console

1. Open NewDot
3. Turn off network
4. Create a workspace
5. Delete the workspace
6. Turn on network
7. Check that the workspace is removed and disappears and it doesn't flash
8. Check for unexpected behaviours and console errors when it comes to network requests and storing data locally

---

1. Check for any unknown behaviour regarding storage/Onyx
2. Check for console errors related to TypeScript, Onyx and Node/NPM

### Offline tests

None needed.

### QA Steps

- [ ] Verify that no errors appear in the JS console

Same as in Tests.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot_20240123_193419_New Expensify Dev

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - iPhone 15 Pro - 2024-01-23 at 19 33 33

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - iPhone 15 Pro - 2024-01-23 at 19 42 31

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-01-23 at 19 30 33

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2024-01-23 at 19 39 42

</details>

### CODE REVIEW

**Review by aimane-chnaif — COMMENTED** [review]
Can you please add Tests / QA Steps of
- #34584
- #34575
- #34586?

**Comment by chrispader** [comment]
@aimane-chnaif yes, but this is only a draft now, everything still WIP.

Still have to figure out all of the problems with the new logic in Onyx

**Comment by VickyStash** [comment]
@chrispader Will it be possible for you to update the Onyx version to `1.0.130` in this PR? That version contains a small [TS fix](https://github.com/Expensify/react-native-onyx/pull/456), which should unblock me in one of mine [migrations](https://github.com/Expensify/App/issues/24926)

**Comment by chrispader** [comment]
> @chrispader Will it be possible for you to update the Onyx version to `1.0.130` in this PR? That version contains a small [TS fix](https://github.com/Expensify/react-native-onyx/pull/456), which should unblock me in one of mine [migrations](https://github.com/Expensify/App/issues/24926)

Yes, i can include the changes from your PR :)

**Comment by chrispader** [comment]
Waiting for https://github.com/Expensify/react-native-onyx/pull/458 to be merged, before we can bump Onyx here...

**Comment by chrispader** [comment]
https://github.com/Expensify/App/pull/33554#issuecomment-1894753320
> Noting there are 3 known issues found after this change for later debugging, we had to revert:
> 
> * [[$500] Login - Can't go back after trying to log in using another tab with invalid code #34584](https://github.com/Expensify/App/issues/34584)
> * [[$500] Sign out - Page gets unresponsive after sign out when login with Abracadabra #34575](https://github.com/Expensify/App/issues/34575)
> * [[$500] Workspace - User redirected to launching expensify screen after selecting free plan on OD #34586](https://github.com/Expensify/App/issues/34586)

The first two regressions are fixed.

Not sure if i can properly test the third regression. I tried following the repro steps, but when i'm already signed into ND with an account and then redirect from OD with a different account, the currently logged in account (in ND) will stay logged in and get redirected to the new workspace page.. Is this intendend?

**Comment by mountiny** [comment]
Making a test build to help with the QA

**Comment by mountiny** [comment]
https://github.com/Expensify/App/actions/runs/7630354768

**Comment by chrispader** [comment]
> @chrispader something is off with the perf test on this branch, seems like they dont even start, seems to work elsewhere

@mountiny updated the node and npm engine version. Hope that fixes it.

**Comment by chrispader** [comment]
> @chrispader something is off with the perf test on this branch, seems like they dont even start, seems to work elsewhere

yea, it won't work because we're using the Node and NPM version from the `BASELINE_BRANCH` (which is `main`)

https://github.com/Expensify/App/blob/d8fb3f45b1852ae01e8a62f16361256ca193fbec/.github/workflows/reassurePerformanceTests.yml#L30-L32

File: .nvmrc
Comment (mountiny) [inline .nvmrc:1]: the reassure tests seem to be failing because of this, why did you update it here?

Comment (chrispader) [reply .nvmrc:1]: yes, because @roryabraham asked me to do so: https://github.com/Expensify/App/issues/12775#issuecomment-1899331180

**Comment by chrispader** [comment]
@mountiny the iOS build fails because of the same error as in https://github.com/Expensify/App/pull/31483.

Adding the `Podfile.lock `diff here too

**Review by neil-marcellini — COMMENTED** [review]
> Not sure if i can properly test the third regression. I tried following the repro steps, but when i'm already signed into ND with an account and then redirect from OD with a different account, the currently logged in account (in ND) will stay logged in and get redirected to the new workspace page.. Is this intendend?

I think we might need an internal employee to test this. Here's what I found when testing.

On production the transition works properly, but the previous tab is broken.

https://github.com/Expensify/App/assets/26260477/dd5867fe-06bf-4574-97d7-68fedcd1ed88


On dev it's the same story, but then things break if you refresh the page after transitioning.


https://github.com/Expensify/App/assets/26260477/9a4e24ba-cbbd-43b7-a92d-c925bce66065


On dev main a similar thing happens, so I don't think this PR is necessarily causing issues with this flow. I think we should proceed and create separate issues to fix the transitions.

**Review by hayata-suenaga — COMMENTED** [review]
the code change looks good to me but I'll wait until0 [this regression](https://github.com/Expensify/App/pull/34801#pullrequestreview-1840096233) is confirmed to have been fixed

File: .nvmrc
Comment (mountiny) [reply .nvmrc:1]: Thanks!

**Comment by chrispader** [comment]
> the code change looks good to me but I'll wait until0 [this regression](https://github.com/Expensify/App/pull/34801#pullrequestreview-1840096233) is confirmed to have been fixed

> On dev main a similar thing happens, so I don't think this PR is necessarily causing issues with this flow. I think we should proceed and create separate issues to fix the transitions.

@hayata-suenaga as far as i understand the comment by @neil-marcellini, this regression is not caused by this PR (that's also what i observed)

should we continue and merge this PR soon? There are some other PRs in E/App waiting for this to be merged :)

**Comment by mountiny** [comment]
https://github.com/Expensify/App/actions/runs/7655884645

**Comment by hayata-suenaga** [comment]
@aimane-chnaif are there any other issues that need to be solved? if not, could you complete the checklist when you have a chance 🙇

**Comment by chrispader** [comment]
> Various github checks are failing

@aimane-chnaif Typecheck pipeline fixed. I think Ressure Performance Tests and `license/snyk` are not related to this PR...

**Comment by chrispader** [comment]
> > Typecheck pipeline fixed
> 
> Screenshot 2024-01-25 at 5 36 11 PM

fixed that already! :)

**Comment by chrispader** [comment]
> > Reassure Performance Tests / perf-tests (pull_request) Failing after 40s
> 
> This is known but after checking log, it's different from others

yes, they are expected to fail, because of the NPM/Node bump

**Comment by chrispader** [comment]
> > > > Reassure Performance Tests / perf-tests (pull_request) Failing after 40s
> 
> > > 
> 
> > > 
> 
> > > This is known but after checking log, it's different from others
> 
> > 
> 
> > yes, they are expected to fail, because of the NPM/Node bump
> 
> 
> 
> Won't this still fail after merge into main?

no, because the the baseline branch also has the same node version.

The reassure pipeline uses the node version from `main` to install the deps. that's why it fails in this PR

**Comment by chrispader** [comment]
> ok can you please check [this bug](https://github.com/Expensify/App/pull/34801#issuecomment-1910437448)? Otherwise looks good!

I've just tested it both on `main` and on this PR. The focus modal is also shown `main` after sign out, so it's not related to this PR.

Also couldn't reproduce the "inifinte loading" on the second tab. By "second tab", do you refer to the one where you put in the link or where you entered you're email.

Is anybody else able to reproduce this bug?

Here's the screen recordings from my machine:

`main`:

https://github.com/Expensify/App/assets/20173411/abb0fa52-71d4-40e2-bebb-b942b0eca5f3

This PR:

https://github.com/Expensify/App/assets/20173411/bdf9e8c5-f997-4a32-ba9d-75b39cc8808d

**Comment by chrispader** [comment]
> first tab: enter email and hit sign in second tab: put in the email link
> second tab: put in the email link

ah ok, that's what i thought. I couldn't repro the infinite loading though..

**Review by aimane-chnaif — APPROVED** [review]

**Review by neil-marcellini — DISMISSED** [review]
Looks good. Thank you all! Let's only revert this if it's proved to be the cause of major regressions.

**Comment by neil-marcellini** [comment]
@chrispader please resolve the merge conflicts

**Comment by chrispader** [comment]
> @chrispader please resolve the merge conflicts

going to do that later today! in ~8hours

**Comment by chrispader** [comment]
> @chrispader please resolve the merge conflicts

@neil-marcellini conflicts resolved!

**Comment by chrispader** [comment]
I'm going to be 100% OOO from 31/01/2024 - 17/03/2024, but i'm available until then. @neil-marcellini maybe we can merge this PR today?

**Review by neil-marcellini — APPROVED** [review]

**Comment by neil-marcellini** [comment]
@chrispader merged!

I merged without Snyk passing because we usually makr Snyk as successful as [noted in Slack here](https://expensify.slack.com/archives/C03TQ48KC/p1706362073693039?thread_ts=1706295170.669819&cid=C03TQ48KC). It says I'm not an admin to mark it as successful but I think this is fine.

**Comment by francoisl** [comment]
I experienced the flashing on one 2 attempts

https://github.com/Expensify/App/assets/2229301/86f4cafc-bdec-4990-a187-4337e8efa5d7

I don't think it's worth blocking the deploy for this though, for now I'll update the PR description so that it doesn't close the parent issue and can be looked into further.

File: src/libs/ReportActionsUtils.ts
Comment (aldo-expensify) [inline src/libs/ReportActionsUtils.ts:1]: I'm getting a type error from this line:

image

image

Does anyone know how to fix it?

Comment (aldo-expensify) [reply src/libs/ReportActionsUtils.ts:1]: Oh, it got fixed after deleting `node_modules` and running `npm install` again

**Pull request merged**
